pkg/upstream/cluster/loadbalancer.go

```diff
@@ -624,3 +625,146 @@ func (lb *reqRoundRobinLoadBalancer) IsExistsHosts(metadata api.MetadataMatchCri
 	return lb.hosts.Size()
 }
+
+const defaultPeakEwmaDuration = 10 * time.Second
+
+type peakEwmaLoadBalancer struct {
+	*EdfLoadBalancer
+	rrLB types.LoadBalancer
+
+	choice          uint32
+	defaultDuration time.Duration
+}
+
+func newPeakEwmaLoadBalancer(info types.ClusterInfo, hosts types.HostSet) types.LoadBalancer {
+	lb := &peakEwmaLoadBalancer{
+		rrLB: rrFactory.newRoundRobinLoadBalancer(info, hosts),
+
+		choice:          defaultChoice,
+		defaultDuration: info.ConnectTimeout() + info.IdleTimeout(),
+	}
+
+	lb.EdfLoadBalancer = newEdfLoadBalancer(info, hosts, lb.unweightedChoose, lb.hostWeight)
+
+	if lb.defaultDuration == 0 {
+		lb.defaultDuration = defaultPeakEwmaDuration
+	}
+
+	return lb
+}
+
+func (lb *peakEwmaLoadBalancer) hostWeight(item WeightItem) float64 {
+	host, ok := item.(types.Host)
+	if !ok {
+		return 1.0
+	}
+
+	stats := host.HostStats()
+
+	duration := stats.UpstreamRequestDurationEWMA.Rate()
+	// None of the active requests returned, or the metrics is disabled
+	if duration == 0 {
+		duration = float64(lb.defaultDuration)
+	}
+
+	return float64(host.Weight()) / (duration * float64(stats.UpstreamRequestActive.Count()+1))
+}
+
+func (lb *peakEwmaLoadBalancer) unweightedChoose(context types.LoadBalancerContext) types.Host {
+	hs := lb.hosts
+
+	var candidate types.Host
+
+	// If `total` is less than or equal to `choice`, we can iterate over all elements directly.
+	if hs.Size() <= int(lb.choice) {
+		candidate = lb.iterateChoose()
+	} else {
+		candidate = lb.randomChoose()
+		if candidate == nil {
+			if log.DefaultLogger.GetLogLevel() >= log.WARN {
+				log.DefaultLogger.Warnf("[lb][PeakEwma] no host chosen after %d choice, fallback to RR", lb.choice)
+			}
+
+			return lb.rrLB.ChooseHost(context)
+		}
+	}
+
+	return candidate
+}
+
+func (lb *peakEwmaLoadBalancer) iterateChoose() types.Host {
+	total := lb.hosts.Size()
+
+	var candidate types.Host
+
+	// avoid always choosing the same host when the score is the same
+	lb.mutex.Lock()
+	idx := lb.rand.Intn(total)
+	lb.mutex.Unlock()
+
+	for i := 0; i < total; i++ {
+		temp := lb.hosts.Get((i + idx) % total)
+		if !temp.Health() {
+			continue
+		}
+
+		if candidate == nil || lb.unweightedPeakEwmaScore(temp) < lb.unweightedPeakEwmaScore(candidate) {
+			candidate = temp
+		}
+	}
+
+	if candidate != nil {
+		if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
+			log.DefaultLogger.Debugf("[lb][PeakEwma] iterate choose host %s with score %d",
+				candidate.AddressString(), lb.unweightedPeakEwmaScore(candidate))
+		}
+	}
+
+	return candidate
+}
+
+// Choose `choice` times and return the best one.
+// See The Power of Two Random Choices: A Survey of Techniques and Results
+//
+//	http://www.eecs.harvard.edu/~michaelm/postscripts/handbook2001.pdf
+func (lb *peakEwmaLoadBalancer) randomChoose() types.Host {
+	total := lb.hosts.Size()
+
+	var candidate types.Host
+
+	for i := 0; i < int(lb.choice); i++ {
+		lb.mutex.Lock()
+		idx := lb.rand.Intn(total)
```

Comment:
> In addition, due to the high cost of achieving random unique selection (the partition selection method will have skew when the score are the same), so still only using rand to ensure that the chosen is as unique as possible

Let's add a comment to tell the readers that we don't ensure uniqueness because of the cost.